lib/fluent/plugin/filter.rb

```diff
@@ -28,24 +28,57 @@ class Filter < Base
       include PluginLoggerMixin
       include PluginHelper::Mixin
 
+      class DuplicatedFilterMethodError < StandardError; end
+
       helpers :event_emitter
 
       def filter(tag, time, record)
         raise NotImplementedError, "BUG: filter plugins MUST implement this method"
       end
 
+      def filter_with_time(tag, time, record)
+        raise NotImplementedError, "BUG: filter plugins MUST implement this method"
+      end
+
       def filter_stream(tag, es)
         new_es = MultiEventStream.new
         es.each do |time, record|
           begin
-            filtered_record = filter(tag, time, record)
-            new_es.add(time, filtered_record) if filtered_record
+            filtered_record = do_filter(tag, time, record)
+            new_es.add(*filtered_record) if filtered_record
           rescue => e
             router.emit_error_event(tag, time, record, e)
           end
         end
         new_es
       end
+
+      private
+
+      def do_filter(tag, time, record)
+        if implemented_method_name == :filter
+          record = filter(tag, time, record)
+          [time, record] if record
+        else
+          filter_with_time(tag, time, record)
+        end
+      end
+
+      def implemented_method_name
```

Comment:
This process (detecting implemented methods) should be done as early as possible. Lazy call (the first time when events arrived) makes data loss if the filter was implemented in invalid way.
IMO it's best to do it in `#initialize`. Raising error in `#initialize` makes fluentd to stop at the time just launched.